client/rest-high-level/src/main/java/org/elasticsearch/client/dataframe/transforms/DataFrameTransformConfig.java

```diff
@@ -0,0 +1,170 @@
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.client.dataframe.transforms;
+
+import org.elasticsearch.client.dataframe.transforms.pivot.PivotConfig;
+import org.elasticsearch.common.ParseField;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.xcontent.ConstructingObjectParser;
+import org.elasticsearch.common.xcontent.ToXContentObject;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.index.query.MatchAllQueryBuilder;
+
+import java.io.IOException;
+import java.util.Objects;
+
+import static org.elasticsearch.common.xcontent.ConstructingObjectParser.constructorArg;
+import static org.elasticsearch.common.xcontent.ConstructingObjectParser.optionalConstructorArg;
+
+public class DataFrameTransformConfig implements ToXContentObject {
+
+    public static final ParseField ID = new ParseField("id");
+    public static final ParseField SOURCE = new ParseField("source");
+    public static final ParseField DEST = new ParseField("dest");
+    public static final ParseField QUERY = new ParseField("query");
+    // types of transforms
+    public static final ParseField PIVOT_TRANSFORM = new ParseField("pivot");
+
+    private final String id;
+    private final String source;
+    private final String dest;
+    private final QueryConfig queryConfig;
+    private final PivotConfig pivotConfig;
+
+    private static final ConstructingObjectParser<DataFrameTransformConfig, String> PARSER =
+            new ConstructingObjectParser<>("data_frame_transform", true,
+                (args) -> {
+                    String id = (String) args[0];
+                    String source = (String) args[1];
+                    String dest = (String) args[2];
+                    // default handling: if the user does not specify a query, we default to match_all
+                    QueryConfig queryConfig = (args[3] == null) ? new QueryConfig(new MatchAllQueryBuilder()) : (QueryConfig) args[3];
+                    PivotConfig pivotConfig = (PivotConfig) args[4];
+                    return new DataFrameTransformConfig(id, source, dest, queryConfig, pivotConfig);
+                });
+
+    static {
+        PARSER.declareString(optionalConstructorArg(), ID);
```

Comment:
I do not think that `ID` is optional. It is required and needs to be non-null. The reason it is an optional arg on the server side is because the user COULD provide it in the URL, but we will make that choice for the user when we transform the rest request.

Reply:
Good spot 👓